keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1598 @@
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Add support to Kubernetes for declaring _quality-of-service_ resources, and
+assigning these to Pods. A quality-of-service (QoS-class) resource is similar
+to other Kubernetes resource types (i.e. native resources such as `cpu` and
+`memory` or extended resources) because you can assign that resource to a
+particular container. However, QoS-class resources are also different from
+those other resources because they are used to assign a _class identifier_,
```

Comment:
Basically, yes. But in this case "thoughtful" resource management is especially a responsibility of the cluster/node administrator. @mikebrow  would you like something about this to be mentioned in here?

Reply:
As I understand it, a QoS-class resource that is unbounded and anyone can claim their usage.  If every pod on a node all requested the same QoS-class resource, the contract implies they all can get it!  This could mean its restricted to compressible resources, but it is not necessarily providing any banding across those resources....

Reply:
yes, I think these additional comments and derek's should be included in some form

Reply:
> a QoS-class resource that is unbounded and anyone can claim their usage. If every pod on a node all requested the same QoS-class resource, the contract implies they all can get it! 

This is true. I think typically the QoS-class resources would be used for throttling, not for providing any guaranteed performance (e.g. in general you can't promise a certain memory bw but you can set limits)

I felt this detail wouldn't belong in the summary. I added to the Proposal section the following paragraph:

*QoS-class resources are unbounded in the way that any number of applications can request the same class (of the same QoS-class resource). QoS-class resources typically represent some throttling mechanism - in this case for example, if all (or most of) applications request the highest-tier class all of these applications get the same level of service. Future work on [access control(#access-control) will provide mechanisms to limit what QoS-class resources (or classes) are available to users.*

WDYT?

Reply:
sgtm

Reply:
> QoS-class resource that is unbounded and anyone can claim their usage. If every pod on a node all requested the same QoS-class resource, the contract implies they all can get it!

@derekwaynecarr @mikebrow regarding the unbounded nature in the latest iteration I now added discussion about "class capacity" in the open questions. So if people want it (and don't get confused) I think it can be added into the APIs relatively easily (i.e. a mechanism to limit the usage of certain classes). Any take on this?

Reply:
@mikebrow as per the comments above marking this as resolved